### PR TITLE
cEP-0015.md: corobo enhancement

### PR DESCRIPTION
**Pull request by nvzard**
Closes https://github.com/coala/cEPs/issues/110

### CODE REVIEW

**Comment by nvzard** [comment]
A lot needs to be added to this cEP at the moment. I'll be adding more details in the coming days.
Please leave reviews, suggestions, and ideas which might help in the development of this project :)

File: cEP-0015.md
Comment (meetmangukiya) [inline cEP-0015.md]: I'd suggest to improve the project description on GSoC website as well, while you can.. maybe after the end of community bonding period.

File: cEP-0015.md
Comment (meetmangukiya) [inline cEP-0015.md]: Define safe mode, what will be disabled, what will be accessible, etc.

File: cEP-0015.md
Comment (meetmangukiya) [inline cEP-0015.md]: Actually all four of this commands are part of LabHub plugin, so not a `List of plugins` per se. No other plugins that are coala specific, are you sure?

Comment (meetmangukiya) [reply cEP-0015.md]: `coatils` for one is coala specific and will remain coala specific, maybe move such specific plugins to a different directory and prevent them from loading for non-coala setup in `Dockerfile`, maybe?

Comment (meetmangukiya) [reply cEP-0015.md]: Other coala specific plugins are:
1. answers  -- we can maybe make this coala independent and make it a generic docs based answering plugin, will require extensive refactoring
2. coala_lowercase_c  - coala specific only, can't adapt
3. explain -- hardcoded coala explanations
4. searchdocs -- ?

**Review by meetmangukiya — DISMISSED** [review]

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: I didn't know we can update the description :o Thank's for informing, its done :+1:

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: @meetmangukiya we can create a separate plugin from `contrib_stats <username>` cmd of `coatils` to show stats of contributions made to the organization. This one can be useful for other organizations?

Comment (nvzard) [reply cEP-0015.md]: And regarding `corobo explain <term>` , I think this one is already generalized.
One can simply just add a template in dir `plugins/templates/explain/` to add new explanations of his choice?
or we can make the user setup a variable `EXPLAIN_DIR` , which will contain locations to the explaination templates?

Comment (jayvdb) [reply cEP-0015.md]: maybe have two sub-directories `explain/generic` and `explain/<org name>` , so this repo can be a depot for explains of other orgs.

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: split lines at 80cpl.

Also run markdownbear on your own file.

Or fix https://github.com/coala/cEPs/issues/125 first

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: For searchDocs plugin, currently API_DOCS and USER_DOCS URLs are constants for coala.
I suggest, we can make it generic by making the orgs setup their custom search URL/URLs in config file.

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: @meetmangukiya I had been thinking about this one.
My idea was that this command will deactivate a list of plugins which we provide in case of emergency situations, but will we need something like this after security hardenings?

Comment (meetmangukiya) [reply cEP-0015.md]: There's already plugin management plugin built into errbot, which can be used to activate and deactivate a specified plugin. If deactivation is the only feature, I don't think we need safe mode.

File: cEP-0015.md
Comment (ishanSrt) [inline cEP-0015.md]: can you explain this a bit?

File: cEP-0015.md
Comment (ishanSrt) [inline cEP-0015.md]: maybe even ban newcomers for a day or two, if they violate too many rules given in the newcomer's guide

File: cEP-0015.md
Comment (ishanSrt) [inline cEP-0015.md]: so like looking through newcomer issues and checking for possibilities like:
* no ones assigned to it
* no one claimed it in the last week
* no one pushed changes or was even active on gitter since the last week, in which case atleast newcomer issues can be reassigned as they don't require more work than that, also pinging newcomers of their unassignment that they havent worked for so and so time on a newcomer issue and it **will** be reassigned, more time gap for low issues
* discarding issues which have users waiting in the queue wanting to get assigned both from gitter and comments on GitHub 

this stuff would be cool af,
would love to see it actually working

File: cEP-0015.md
Comment (ishanSrt) [inline cEP-0015.md]: would be great if you can stuff some more implementation details into this 😉

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: Won't that be too harsh :P ?
The developers/maintainers will invite them and assign the issue only after they find one to work on.

Comment (ishanSrt) [reply cEP-0015.md]: thats why just for a day or two, asking them nicely to again go through the newcomer's guide and some git tutorials and such

Comment (ishanSrt) [reply cEP-0015.md]: anyways its your call

Comment (ishanSrt) [reply cEP-0015.md]: if they are doing so many things wrong, means they have taken this lightly and haven't gone through the provided content. also asking too many googlable questions may count to that

File: cEP-0015.md
Comment (jayvdb) [reply cEP-0015.md]: Is there an issue about this feature?

If so, please mention it.

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: > and extend testing ...

What changes to errbot do you have in mind, *other* than testing?
Details please, or dont mention it here.

Fixing their testing framework, and implementing those changes, is priority no 1 for your project.  corobo becomes abandonware if you dont get testing working.  And likely your project is failed after coding period 1 if you haven made significant inroads into solving that.

Need to see a lot more details about that one specifically.

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: No there is no issue yet, but I'll create one and mention it here.

Comment (ishanSrt) [reply cEP-0015.md]: not about this feature, kind of comparing this cEP in general to others

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: I agree with your point but that might scare them away from contributing :P 
We can definitely add a command to warn them about not following the guidelines.
I think we'll need some discussion on this point.

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: this isnt security.
the only way that helps security is if being in a gitter room gives the user access to sensitive/restricted operations, which sounds like a very bad idea as we cant control people joining the room with lots of accounts, either themselves or getting their friends to help coordinate an attack.

this is access control, but the goal of it is to prevent spamming.

File: cEP-0015.md
Comment (ishanSrt) [reply cEP-0015.md]: as I said its completely your call on that 👍

Comment (nvzard) [reply cEP-0015.md]: We can add a scarscatic explaination for `corobo explain google` :P

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: Firstly `We` doesnt belong in a spec/proposal.

But more importantly, this starts the proposal re security on the wrong foot.

Here you are dictating implementation details in the `Why?`.  That eliminates the need for requirements gathering, and usually ends up with failure.

The sole purpose of corobo is to empower coala members to do things according to the orgs defined roles, giving them *more* power than is possible with GitHub's and GitLab's limited (and often bizarre) access control mechanisms.

As an extremely open org (and sadly you havent experienced it as much as the previous GSoC students due to the aforesaid problems), any access control is inherently bad and must be justified as a necessary evil.

corobo is a balancing act of giving org members rights to do whatever they need to do, vs preventing them from causing mayhem.

We havent got that balancing act right yet, hence this project ;-)

GitHub team memberships are a means to an end.
The most important requirement is allocation of issues to people, and for people's first issue, that means we have no idea (zero trust) in the person being given org rights.
However team(org) membership comes with serious problems.  Foremost is that any member of the org has rights given by GitHub which we probably dont want them have, and have little counter-measures or *tested* procedures to deal with.  This is where the testability aspect comes in.  99.99% of the time we dont need these counter-measures.  But the one day we do need to react, we are going to be very reliable on counter-measures which havent been used for months or years.  A reliable test-suite for counter-measures gives us confidence that we can grant newcomers rights even if we dont yet trust them, because we are able to react effectively enough to halt bad actors soon enough that the cost/benefit is stacked in favour of given the rights.

Read through https://help.github.com/articles/permission-levels-for-an-organization/ and https://help.github.com/articles/repository-permission-levels-for-an-organization/ to see how screwed we are if a newcomer wants to be disruptive.

One of the recently found problems is that people can do a merge on their own repos which closes issues in the org, if they use `Closes coala/coala#123`.  Maybe that is only org members?  We havent checked that yet, but that would be one of the more extremely disruptive problems. (an abusing person might batch close all issues..)

See https://gitlab.com/coala/roles/ for a possible way to avoid GitHub team membership for newcomers who havent yet been found to be trustable.  See https://docs.gitlab.com/ee/user/permissions.html for a much saner set of minimums.  But still not granular enough, and contains a few concerning permissions.  This would mean we wouldnt be able to assign GitHub issues to them.

Another option is to use another org like 'coala-newcomers' for the newcomers, so they have no rights in the 'coala' org, but then we cant assign issues to them.

Another option is to not use GitHub or GitLab for users until they have been vetted, but use a user list stored somewhere else which only has the access which corobo has configured for them, and no implicit rights given by the repo hoster.  Again, this would mean we cant assign issues to them on GitHub.

Notice a theme here ... assigning issues is our process problem wrt GitHub.  The GitMate project has a few possible solutions which havent been road tested, and they wont be perfect.  The corobo project needs to also have solutions for this problem, and again they will be limited workarounds.  It is a significant enough problem, and coala isnt the only org with this problem, that having multiple different solutions are needed, and may need to be used together to achieve the right balance, or at least different orgs will want to use different solutions, and coala's mission includes building tools that other orgs can use, so offering multiple solutions empowers those orgs to find the one that suits them best...

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: @meetmangukiya creating generic docs based answering plugin will require changes in the `answers` microservice itself as the docs are parsed there. How can we go about making this generic for every organization o.O ?

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: https://help.github.com/articles/closing-issues-using-keywords/
It says `Including Closes example_user/example_repo#76 will close the referenced issue in that repository, provided you have push access to that repository.` 
Only maintainers/owners have push/write access, it won't be possible for newcomers to close issues?
>people can do a merge on their own repos which closes issues in the org, if they use Closes coala/coala#123. Maybe that is only org members? We havent checked that yet, but that would be one of the more extremely disruptive problems

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: Okay, I'll reframe the structure of this cEP and count this in as a measure to prevent spamming than security.

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: @ishanSrt  Thanks for the suggestion but I guess this would be GitMate type of stuff and might not be possible to write an Errbot plugin for this.
@Vamshi99 what do you say about this suggestion?

File: cEP-0015.md
Comment (gbin) [inline cEP-0015.md]: We started to brainstorm about it with @nvzard. We are thinking about extending a testbot with a feature to be able to inject mocks in the plugins.
This would look something like this (but needs to be tried out to be sure it is user friendly):
```python
import pytest

fakeObj1 = FakeObj()  # your mock
extra_plugin_dir = ...  # path to your plugin to test (already implemented)
inject_in_plugins = { 'PluginName': {'field_name1': fakeObj1,
                                                         'field_name2': fakeObj2}}
def test_mycommand(testbot):
    assert 'something' in testbot.exec_command('!my_command')
```
Another way or complementary way might be (if we want to change mocks at every test)
```python
def test_mycommand(testbot):
    testbot.inject_mock('PluginName', 'field_name1', fakeObj1)
    assert 'something' in testbot.exec_command('!my_command')
```

(Then it will have to so some magic to inject a property on the plugin object to intercept and discard the assignments from the plugin itself.)

File: cEP-0015.md
Comment (gbin) [inline cEP-0015.md:151]: Another thing to do here is to make the plugins use activate and not `__init__` for anything requiring the plugin to be configured and activated.

For example in labhub.py:
```python
    def __init__(self, bot, name=None):
        super().__init__(bot, name)

        teams = dict()
        try:
            gh = github3.login(token=os.environ.get('GH_TOKEN'))
            assert gh is not None
        except AssertionError:
            self.log.error('Cannot create github object, please check GH_TOKEN')
        else:
            self.GH3_ORG = gh.organization(self.GH_ORG_NAME)
            for team in self.GH3_ORG.teams():
                teams[team.name] = team

        self._teams = teams

        self.IGH = GitHub(GitHubToken(os.environ.get('GH_TOKEN')))
        self.IGL = GitLab(GitLabPrivateToken(os.environ.get('GL_TOKEN')))

        self.REPOS = dict()
```

File: cEP-0015.md
Comment (meetmangukiya) [reply cEP-0015.md]: This looks good :D :+1:

File: cEP-0015.md
Comment (meetmangukiya) [reply cEP-0015.md:151]: I don't quite remember why I made this choice, but I think it was the API calls, we didn't want to make a ton of API requests every time the plugin is deactivated and activated again.. Since errbot just instantiates the plugin once, it sounded like a better choice.. not the best choice to make I guess :sweat_smile: 

What are the downsides of setting the attribs in the `__init__`?

Comment (gbin) [reply cEP-0015.md:151]: The issue of `__init__` is that you do that at load time, if there is any issue the plugin doesn't even show up.
Also the cycle lifecycle is `__init__` -> configured -> enabled
So if you have state in your config for your connection like passwords etc you can't get those from `__init__`.

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: I suggest the `inject_mock` takes a dict of mock objects as parameter so that we can mock more that one field for a bot command and the user is able to inject the dict for every command_test:
```
def test_mycommand(testbot):
    mock_config = {'PluginName': {'field_name1': fakeObj1,
                                   'field_name2': fakeObj2}}
    testbot.config_mocking(mock_config)
    assert 'something' in testbot.exec_command('!my_command')
```

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md:100]: I created a [bunch of issues](https://github.com/coala/corobo/issues?utf8=%E2%9C%93&q=is%3Aopen+get_configuration_template+) about this, including some not mentioned below, but when I came to do it myself I found a problem is that we need configuration to be pre-loaded.

It isnt appropriate for a bot admin to need to run `!config` every time the bot starts up, for every plugin.
It needs to reload the previous saved state, or it needs to start using config from a static config file.

Maybe I missed something important.  See https://github.com/coala/corobo/issues/380#issuecomment-391575383 for a quick description of the problem and my workaround.

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: incorrect indentation

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: Wouldnt it use the GitHub API using `IGitt`?

If it is necessary to use a scraper, you will need to explain why a proper API can not be used.

File: cEP-0015.md
Comment (jayvdb) [reply cEP-0015.md]: https://github.com/coala/corobo/issues/528 has been created.

IMO this cEP, and probably even this project, should not have new features which are not related to "security, tests and configurability".  Your project summary in GSoC doesnt mention new features.  Your GSoC proposal did mention adding this feature, but it definitely isnt relevant to the stated scope of this cEP.

It might be still included in your milestone tasks, but it may be dropped if the project team agrees that other pending work is more important. e.g. one of the critical components of this project takes more time than expected. This feature is useless if the corobo hardening isnt completed as best as is possible.

Anyone could add new features like this.  It doesnt require a GSoC student to do it.  GCI students last year built bigger more complicated plugins in less than three days.  If you dont build it this GSoC, no drama .. a newcomer or GCI student will build it by the end of the year.

Also I feel that it would be inappropriate to build more GitHub-only features into corobo.  corobo should exclusively use IGitt, and all plugins should work in all hosters supported by IGitt, unless it is functionality which is only supported by one hoster and it is really critical that it is available in corobo.

fwiw, another [2018 GSoC project](https://github.com/coala/projects/commit/f5ca3b597f9693fd82e7e71072deda63abed701a) will be providing a GitHub+GitLab issue search tool.

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md]: Okay, I'll remove this section since it is out of scope for this project :)

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md:100]: @jayvdb I understand the problem but the configuration_template is there to avoid the need for admins to configure settings in some kind of configuration file, allowing configuration to happen directly through chat commands.

I guess configuring keys with empty `{TOKEN: ''}` values if they exist from the global config file will deal with this type of situation.

Comment (nvzard) [reply cEP-0015.md:100]: @jayvdb  according to @gbin configurations are stored in ErrBot and last saved configuration are loaded after restarting the bot.

Comment (jayvdb) [reply cEP-0015.md:100]: there is still the problem of first time the bot is started.
the configuration of the corobo plugins needs to initialised somehow.

Comment (gbin-argo) [reply cEP-0015.md:100]: Sure you here you have several choices: either a restore from a backup or you can insert values into storage from errbot's command line before startup. Errbot is not a stateless system but it can use existing external DBs as storage if you wish (SQL, redis, gcd etc.).

Comment (jayvdb) [reply cEP-0015.md:100]: We should choose a method, and document it here.

corobo is run in a docker, without any db, so a good choice would be a light-weight load method.

The reason why I wrote my patch above like that was so that the initial values of the plugin configuration was the same code as the plugin configuration code.

File: cEP-0015.md
Comment (nvzard) [inline cEP-0015.md]: Will be removed on next push.

File: cEP-0015.md
Comment (saksham189) [inline cEP-0015.md]: Errbot

Comment (saksham189) [reply cEP-0015.md]: capital E in Errbot

File: cEP-0015.md
Comment (meetmangukiya) [reply cEP-0015.md:100]: I agree with @jayvdb on this one, reconfiguring again and again using the chat via a admin is not preferable. Also, it is worth noting that corobo is continuously deployed on each commit, hence, the usage of previous config stored in the db is not useful either. We need a way to preload the configuration.

**Review by meetmangukiya — COMMENTED** [review]
Most of this looks pretty good, lets figure out the preloading of configuration part, and also use `py` for syntax highlighting the code.

File: cEP-0015.md
Comment (nvzard) [reply cEP-0015.md:100]: Can't we configure the plugins through [provisioning](http://errbot.io/en/latest/user_guide/provisioning.html#writing-values) before the bot is deployed on docker?

Comment (jayvdb) [reply cEP-0015.md:100]: try it and tell us if it works.
This is your project.

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: Who creates this 'dictionary' to begin with?

Will corobo/errbot write this file?

Have you tried this with a simple plugin which uses bot plugin config?
Or does it only work with bot plugin storage?

File: cEP-0015.md
Comment (jayvdb) [reply cEP-0015.md:100]: A new section 'Configuration of corobo' has been added below to include some provisioning of configuration.

That is confusing as this is where you say how people will configure corobo. After reading this extra section I still have no idea how you envisage this working.  I use !config and somehow i get a config file which is somehow loaded into corobo the next deploy....

Comment (nvzard) [reply cEP-0015.md:100]: We had a discussion on this point. 
It is possible to configure the plugins before deploying the bot.
We came to the conclusion that a python file containing the configurations will be mounted to the container from the host machine and we'll run a script from inside the container that will use the provisioning commands to configure the plugins from the configurations inside the python file before the bot is deployed. So that configs are stored in the storage.

Comment (jayvdb) [reply cEP-0015.md:100]: Please help me out here.

http://errbot.io/en/latest/user_guide/provisioning.html

says

> Plugins can be configured by talking to the bot with:
> !plugin config my_plugin {'key': 'value'}
> Also plugins can store values in the storage as key value pairs.
> ...
> echo "{'FACTOID': {'errbot': 'awesome'}}" | errbot --storage-merge Factoid

Can the *same* config variable by configured by `!config` and by `errbot --storage-merge...` ?

Are they a unified system?  If so, show a worked example using the same config variables, e.g. `GH_TEAM_NAMES` and `GH_TOKEN` and `GH_ORG_NAME`, and then also state unambiguously that the bot will react to `!config` changes of those critical config variables, and reload correctly with the new settings.  Also specifically wrt labhub at startup for repo forks, who do not have sufficient access for the org in `GH_ORG_NAME`, the plugin needs to gracefully handle failures in those config values during startup, such as tokens with incorrect scopes.  There are lots of issues about this, and it is limiting the ability of newcomers to get involved in this project.

I havent seen a bot plugin which has the same config variable in storage and accessible via `!config`, so pardon my skepticism and growing impatience with this problem, which was raised here *4 weeks* ago.  Dumping into the cEP examples which show different mechanisms being used doesnt describe a solution ... it just shows errbot has lots of available options ... and pretends they can be used in a unified way when they can not.

If the can not be used in a unified way, this cEP is confusing because it is talking about two different ways of configuring the bot and not clarifying which types of config settings will be done in each of the two ways.

One possible way to solve this is to consider how it might be possible to have the same config variable in storage and accessible via `!config`.  It might be a hack in corobo, like https://github.com/coala/corobo/issues/380#issuecomment-391575383  mentioned 4 weeks ago, or it might be an enhancement to errbot if upstream are interested in solving this problem.

**Review by jayvdb — CHANGES_REQUESTED** [review]

**Comment by gbin** [comment]
The configs you can set with `!plugin config` can be set with the command
line under the key 'core' like the documentation shows. They are completely
equivalent so I think it is relevant for your requirement for provisioning
your plugins ahead of time.

Plugins have a place to validate their config is well formed by overriding
`check_configuration` and will send to the bot admin a message explaining
why a plugin could not start in case of failure.

File: cEP-0015.md
Comment (gbin-argo) [reply cEP-0015.md:100]: The configs you can set with `!plugin config` can be set with the command line under the key 'core' like the documentation shows. They are completely equivalent so I think it is relevant for your requirement for provisioning your plugins ahead of time.

Plugins have a place to validate that their config is well formed by overriding `check_configuration` and will send to the bot admin a message explaining why a plugin could not start in case of failure.

Comment (jayvdb) [reply cEP-0015.md:100]: If they can be used together, that is great.

Below this cEP uses

```
errbot --storage-merge LabHub {'labhub': {'GH_ORG_NAME': 'coala'}}
```

Is that incorrect?  Should `LabHub` be `core`?

As mentioned above, I would want to see at least the same config variables in the `!config` example above and the `errbot --storage-merge` example below.

It would also be desirable to have a list of the config variables that this project will be implementing, so we can see and discuss them, and verify that the planned work is sufficient for the objective of the project.

Comment (nvzard) [reply cEP-0015.md:100]: `errbot --storage-get core` will return configs of all the loaded plugins including LabHub,
 while `errbot --storage-get LabHub` will return config only for LabHub plugin.
Similar for `--storage-merge` and `--storage-set` flags,

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: strange indentation happening here

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: line length

File: cEP-0015.md
Comment (jayvdb) [inline cEP-0015.md]: GH_ORG_NAME = OPENHUB_TOKEN ? ;-)

**Review by jayvdb — CHANGES_REQUESTED** [review]
Minor issues

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Review by jayvdb — APPROVED** [review]

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
ack https://github.com/coala/cEPs/pull/113/commits/d20be0e33dfb9f026af3bc3b30b11f2b2609e99e

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Pull request merged**
